lib/filter_cli.c

```diff
@@ -0,0 +1,1814 @@
+			 "%s/entry[sequence='%s']", xpath, seq_str);
+
+	nb_cli_enqueue_change(vty, xpath_entry, NB_OP_CREATE, NULL);
+
+	snprintf(xpath_value, sizeof(xpath_value), "%s/action", xpath_entry);
+	nb_cli_enqueue_change(vty, xpath_value, NB_OP_MODIFY, action);
+
+	if (prefix_str != NULL) {
+		snprintf(xpath_value, sizeof(xpath_value), "%s/ipv4-prefix",
+			 xpath_entry);
+		nb_cli_enqueue_change(vty, xpath_value, NB_OP_MODIFY,
+				      prefix_str);
+
+		snprintf(xpath_value, sizeof(xpath_value),
+			 "%s/ipv4-exact-match", xpath_entry);
+		nb_cli_enqueue_change(vty, xpath_value, NB_OP_MODIFY,
+				      exact ? "true" : "false");
+	} else {
+		snprintf(xpath_value, sizeof(xpath_value), "%s/any",
+			 xpath_entry);
+		nb_cli_enqueue_change(vty, xpath_value, NB_OP_CREATE, NULL);
+	}
+
+	return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFPY(
+	no_access_list, no_access_list_cmd,
+	"no access-list WORD$name [seq (1-4294967295)$seq] <deny|permit>$action <A.B.C.D/M$prefix [exact-match$exact]|any>",
+	NO_STR
+	ACCESS_LIST_STR
+	ACCESS_LIST_ZEBRA_STR
+	ACCESS_LIST_SEQ_STR
+	ACCESS_LIST_ACTION_STR
+	"Prefix to match. e.g. 10.0.0.0/8\n"
+	"Exact match of the prefixes\n"
+	"Match any IPv4\n")
+{
+	struct access_list *acl;
+	struct lyd_node *dnode;
+	int64_t sseq;
+	struct prefix pany;
+	char xpath[XPATH_MAXLEN];
+	char xpath_entry[XPATH_MAXLEN + 32];
+
+	/* If the user provided sequence number, then just go for it. */
+	if (seq_str != NULL) {
+		snprintf(
+			xpath, sizeof(xpath),
+			"/frr-filter:lib/access-list[type='ipv4'][name='%s']/entry[sequence='%s']",
+			name, seq_str);
+		nb_cli_enqueue_change(vty, xpath, NB_OP_DESTROY, NULL);
+		return nb_cli_apply_changes(vty, NULL);
+	}
+
+	/* Otherwise, to keep compatibility, we need to figure it out. */
+	snprintf(xpath, sizeof(xpath),
+		 "/frr-filter:lib/access-list[type='ipv4'][name='%s']", name);
+
+	/* Access-list must exist before entries. */
+	if (yang_dnode_exists(running_config->dnode, xpath) == false)
+		return CMD_WARNING;
+
+	/* Use access-list data structure to fetch sequence. */
+	dnode = yang_dnode_get(running_config->dnode, xpath);
+	acl = nb_running_get_entry(dnode, NULL, true);
+	if (prefix == NULL) {
+		memset(&pany, 0, sizeof(pany));
+		pany.family = AF_INET;
+		sseq = acl_zebra_get_seq(acl, action, &pany, exact);
+	} else
+		sseq = acl_zebra_get_seq(acl, action, (struct prefix *)prefix,
+					 exact);
+	if (sseq == -1)
+		return CMD_WARNING;
+
+	snprintf(xpath_entry, sizeof(xpath_entry),
+		 "%s/entry[sequence='%" PRId64 "']", xpath, sseq);
+	nb_cli_enqueue_change(vty, xpath_entry, NB_OP_DESTROY, NULL);
+
+	return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFPY(
+	no_access_list_all, no_access_list_all_cmd,
+	"no access-list WORD$name",
+	NO_STR
+	ACCESS_LIST_STR
+	ACCESS_LIST_ZEBRA_STR)
+{
+	char xpath[XPATH_MAXLEN];
+
+	snprintf(xpath, sizeof(xpath),
+		 "/frr-filter:lib/access-list[type='ipv4'][name='%s']", name);
+	nb_cli_enqueue_change(vty, xpath, NB_OP_DESTROY, NULL);
```

Comment:
This isn't working in some cases. Example:
```
access-list 10 seq 5 permit 1.1.1.1/32
no access-list 10
```

Doing a `sh run` will still show that ACL. If I use a word for the ACL name instead if `10` the command can be removed normally.

I suppose we need to use `access-list-legacy` instead of `access-list` in some cases (numeric names that fall into certain ranges)? This applies to other commands like "no access-list WORD remark" as well.

Reply:
I'd say: avoid naming your zebra-style lists with numbers only.

The cisco-style lists have these CLI ambiguities even before northbound, it is really hard to tell from tabbing/auto completing in CLI when you are using Cisco vs Zebra if the name contains numbers.

Reply:
Let me clarify the ambiguity here.

Zebra style access list command format is: `[no] access-list WORD ...`.
Cisco style access list command format is `[no] access-list <(1-99)|(100-199)|(1300-1999)|(2000-2699)>`

This command creates a Zebra-style access list:
`access-list 10 seq 5 permit 1.1.1.1/32`

Because of the `permit 1.1.1.1/32` part, if you wanted to use Cisco style you should have created it like this:
`access-list 10 seq 5 permit 1.1.1.1 255.255.255.255`

My recommendation is: use numbers only naming for Cisco Style ONLY and words mixed with other characters for zebra style lists.